README.md
# BaiduTieBaOffline

#发现有些贴子已经不能下载了，原因是 https://tieba.baidu.com/p/totalComment?tid= 可能返回结果不正确
python贴吧爬虫玩具，把整个贴吧html（除了我认为不需要的，比如广告＝。＝，不愧是某厂，这东西一点都不少）都爬下来，保留基本贴吧页面样式

分为两个部分，爬取贴吧帖子列表和爬取单个帖子所有楼层，保存为本地html包括资源，提供两个模版，基本照着贴吧的样式

帖子列表由于是bigpipe动态加载的，所以用了selenium＋PhantomJS，请保证这两个东西事先装好，跑程序只用传入贴吧首页地址即可，跑完之后会生成贴吧名字的文件夹，里面有1～n页的帖子列表页内容（只是列表页）

帖子详情页由于没有用bigpipe所以直接用urllib请求的，但是需要登录cookie所以这个自己搞个账号自己抓包吧，模拟登陆目前没做，跑程序需要贴吧id和帖子id，可以在帖子第一页源代码里找，或者列表页模块生成出列表页之后也会爬出这些数据，程序跑完之后会生成帖子id的文件夹，里面有帖子第1～n页的所有内容

你也可以在保证selenium＋PhantomJS和登陆cookie的情况下直接传一个贴吧首页地址，它能把整个贴吧爬下来，之后大概长这样(杰钢队长吧为例):

列表页:
页头
![image](https://github.com/ytinrete/BaiduTieBaOffline/raw/master/show_pic/l1.png)
页尾巴
![image](https://github.com/ytinrete/BaiduTieBaOffline/raw/master/show_pic/l2.png)
第二页头
![image](https://github.com/ytinrete/BaiduTieBaOffline/raw/master/show_pic/l3.png)

详情页:
头部
![image](https://github.com/ytinrete/BaiduTieBaOffline/raw/master/show_pic/d1.png)
评论也被爬下来的，但是没有分页，直接有多少贴多少这样
![image](https://github.com/ytinrete/BaiduTieBaOffline/raw/master/show_pic/d2.png)

(本来想拿一两页样本放git上不过两页就100+M。。。,因为包括所有帖子内容,这么大不奇怪,所以还是算了)

## (我只是随便选取的页面,要是你是其中某人而不想被帖在这些图里,给我发邮件立即就删)

无论是列表模块还是详情模块都能根据已知的进度继续爬,所以要是出错了删掉最新的那一页,再继续运行就好,对于要爬整个贴吧还有问题,我请求的时候是不带时间参数的,若是很活跃的贴吧估计帖子数可能会有出入,重复,少掉等这些都不好说,这一部分要优化还得做一些工作,但是我没有这样的需求呀

上面爬下来的都是带资源和做好下一页跳转修正的，也就是基本可以当作拿到整个离线内容了，爬完之后用浏览器就能在本地看帖了，然而并没有什么意义，写这个只是拿来保存我的个人贴吧而已XD

这个玩具爬比较简单的小贴吧应该是没有多大问题（比如说我自己的那个），其他可能会出问题，特别是里面抓取数据代码写的不够灵活，很多地方可以用多线程去做也没做，没优化，因为对我而言也够用了，之后百度api改了那也没办法，总之我已经玩够了2333，不过要真有人用这个遇到问题随时欢迎讨论 --2016.8.25

special thanks:
曾经一起度过美好时光的贴吧吧友们～

---
图片内容可使用tiebapic.baidu.com/forum/pic/item/或者imgsrc.baidu.com/forum/pic/item/+图片名字 去查询

使用的第三方库或软件：

python3 -m pip install selenium==2.53.6（这个版本很低，如果以后换浏览器记得升级下）

PhantomJS（https://github.com/ariya/phantomjs） ，下载后放在python根目录的Scripts文件夹里。这是一个不显示内容的浏览器

python3 -m pip install beautifulsoup4

python3 -m pip install lxml

python3 -m pip install --upgrade pip

//pip3 install --upgrade pip

python版本：Python 3.5.1/3.6.8

单个贴子删除base_info.json文件可以继续更新（注意备份，防止因为原贴已被删除导致丢失了 ）

获取楼层失败可以通过删除失败的楼层文件夹来解决

要爬取单个贴子的话，修改get_single_thread.py里的get_single_thread('贴子id（数字，贴子链接后面的数字）','','','1')即可（支持多行）

要爬取整个贴吧的话，修改get_forum_all.py里的my_url = "https://tieba.baidu.com/f?ie=utf-8&kw=xxxxxxxxxxxx" 即可（支持多行）

2020-2-25




CHANGELOG.md
# 更新日志

## 2021

### 3-4 v0.06
- 支持直接复制粘贴贴子链接爬取了

## 2020

### 5-31 v0.03
- 修复当get_single_thread.py单独使用时，3个以上贴子被删除会导致爬虫停止的bug
- 尝试缓解因为网页内容未完全加载全而导致的爬虫出错

### 3-4 v0.02

- 现在get_single_thread.py单独使用时，可以直接复制粘贴链接了，稍微方便一点

### 3-2 v0.01

- 给get_single_thread.py增加链接超时5秒跳出
- 给get_forum_list.py增加修改header？

get_forum_all.py
import get_forum_list
import get_single_thread
import os
import sys
import shutil  # 文件操作
import logging  # 引入logging模块
import signal
import json
import time
import random
import threading
import re

from urllib.parse import unquote
from selenium import webdriver
from datetime import datetime
from bs4 import BeautifulSoup
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities

# from selenium import webdriver#没有使用谷歌的chromedriver，因为运行起来意外的慢,而且这个要求本机必须安装有注册的谷歌浏览器
#from selenium.webdriver.chrome.options import Options

# path =r'D:\python3.6.8\Scripts\chromedriver.exe'#有附带chromedriver.exe，但要注意selenium的版本号，这里为了支持PhantomJS.exe,压低版本了(2.53.6)
#chrome_options = Options()
# chrome_options.add_argument('--headless')

cookie = ""

# 第一步，创建一个logger
logger = logging.getLogger()
logger.setLevel(logging.INFO)  # Log等级总开关
# 第二步，创建一个handler，用于写入日志文件
rq = time.strftime('%Y%m%d%H%M', time.localtime(time.time()))
if os.path.exists("./logs") == False:
    os.makedirs("./logs")  # 创建logs文件夹用来存放日志
log_path = "./logs/"
log_name = log_path + rq + '.log'
logfile = log_name
fh = logging.FileHandler(logfile, mode='w')
fh.setLevel(logging.DEBUG)  # 输出到file的log等级的开关
# 第三步，定义handler的输出格式
formatter = logging.Formatter(
    "%(asctime)s - %(filename)s[line:%(lineno)d] - %(levelname)s: %(message)s")
fh.setFormatter(formatter)
# 第四步，将logger添加到handler里面
logger.addHandler(fh)
# 日志
#logger.debug('this is a logger debug message')
#logger.info('this is a logger info message')
#logger.warning('this is a logger warning message')
#logger.error('this is a logger error message')
#logger.critical('this is a logger critical message')
# python中logging日志模块详解
# https://www.cnblogs.com/xianyulouie/p/11041777.html

def getSource():
    headers = {
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Content-Encoding': 'gzip,deflate,br',
        'Content-Type': 'text/html; charset=utf-8',
        'User-Agent':'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36',
        'Cookie':''  
    }
    #使用copy()防止修改原代码定义dict
    cap = DesiredCapabilities.PHANTOMJS.copy() 
 
    for key, value in headers.items():
        cap['phantomjs.page.customHeaders.{}'.format(key)] = value
 
    # 不载入图片，爬页面速度会快很多
    cap["phantomjs.page.settings.loadImages"] = False
 
    driver = webdriver.PhantomJS(desired_capabilities=cap)#注意selenium的版本，高版本才支持chromedriver.exe.这里selenium==2.53.6
    return driver

def get_forum_list_call(url):
    #get_forum_list.browser = webdriver.Chrome(executable_path=path, options=chrome_options)
    get_forum_list.browser = getSource()
    get_forum_list.get_forum_list(url)
    get_forum_list.browser.service.process.send_signal(signal.SIGTERM)
    get_forum_list.browser.quit()


def kill_fake_thread(page_path, page, tid):
    with open(page_path, 'r', encoding='utf-8') as f:
        html_tree = BeautifulSoup(f, 'lxml')
    target_node = html_tree.find('a', href=str(page) + '_thread/' + tid + '/1.html')
    if target_node:
        for parent in target_node.parents:
            try:
                if 'j_thread_list' in parent['class']:
                    print('found you!')
                    parent.decompose()
                    found = True
                    break
            except BaseException:
                print('continue find!')
        if found:
            os.remove(page_path)
            with open(page_path, 'w', encoding='utf8') as f:
                f.write(str(html_tree))
            print('kill success!')
        else:
            print('!!!--error--!!! kill fail')
    else:
        print('!!!--error--!!! kill fail')


def get_threads_from_forum_page(url, cookie):
    get_single_thread.cookie = cookie
    info_file_name = unquote(
        unquote(get_forum_list.get_info_from_query(url, 'kw')))
    if os.path.exists(info_file_name + '/forum_base_info.json'):
        with open(info_file_name + '/forum_base_info.json', 'r', encoding='utf-8') as base_info_file:
            info = json.load(base_info_file)

        for page in range(1, info['total_page'] + 1):
            if os.path.exists(info_file_name + '/' + str(page) + '_thread/tid_info.json'):
                with open(info_file_name + '/' + str(page) + '_thread/tid_info.json', 'r',encoding='utf-8') as thread_info_file:
                    tid_info = json.load(thread_info_file)
                fid = tid_info['fid']
                for tid in tid_info['tid_list']:
                    if not os.path.exists(info_file_name + '/' + str(page) + '_thread/' + tid + '/1.html'):
                        try:
                            print('downloading thread ' +
                                  tid + ' in page' + str(page))
                            has_fake_thread = get_single_thread.get_single_thread(tid, fid, info['title'], str(page),True)
                            if has_fake_thread:
                                kill_fake_thread(info_file_name + '/' + str(page) + '.html', page, tid)
                            else:
                                shutil.copytree(tid, info_file_name + '/' + str(page) + '_thread/' + tid)
                                shutil.rmtree(tid)
                        except BaseException:
                            print('!!!--error--!!! download error tid:' + tid)
def start(url):
    print(datetime.now())
    get_forum_list_call(url)
    get_threads_from_forum_page(url, cookie)
    print(datetime.now())
    # Python3 正则表达式 | 菜鸟教程
    # https://www.runoob.com/python3/python3-reg-expressions.html
    #数字正则表达式 - miyaye - 博客园
    # https://www.cnblogs.com/webskill/p/7422876.html


def usejson():
    # 读取JSON配置文件
    filename = "./tieba.json"
    jsontemp = None
    f_obj = None
    try:
        f_obj = open(filename, encoding="utf-8")
        jsontemp = json.load(f_obj)
        f_obj.close()
    except Exception as err:  # FileExistsError or OSError:
        print("读取配置文件失败！ "+str(err))
        f_obj.close()
        exit()
    tiebalists = jsontemp['tieba']
    for x in tiebalists:
        if x[2] == True:
            print('链接:'+x[0]+',标题'+x[1])
            start(x[0])
        else:
            print('链接:'+x[0]+","+str(x[1])+",该贴吧不更新！")
        time.sleep(random.choice(range(5, 10)))
    '''
    https://www.cnblogs.com/lpdeboke/p/11414254.html
    python中json的基本使用
    https://jingyan.baidu.com/article/c74d6000d138fb0f6b595d45.html
    如何使用python的json模块从json文件读取数据听语音
    https://blog.csdn.net/weixin_41931602/article/details/80557416
    python ： 'gbk' codec can't decode byte 0xbe in position 18: illegal multibyte sequenc
    '''


def starttimer():
    global timer
    # timer.cancel()
    # print("2233")
    usejson()
    timeoutx = 7200+random.choice(range(1, 10))
    print("延迟："+str(timeoutx)+"s,下次运行时间："+str(time.strftime("%Y-%m-%d %H:%M:%S %a",time.localtime(time.time()+timeoutx))))#接收时间戳（1970纪元后经过的浮点秒数）并返回当地时间下的时间元组t（t.tm_isdst可取0或1，取决于当地当时是不是夏令时）。
    timer = threading.Timer(timeoutx, starttimer)
    timer.start()
    '''
    格式化日期
    我们可以使用 time 模块的 strftime 方法来格式化日期，：

    time.strftime(format[, t])
    实例演示：

    实例(Python 2.0+)
    #!/usr/bin/python
    # -*- coding: UTF-8 -*-
    
    import time
    
    # 格式化成2016-03-20 11:45:39形式
    print time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()) 
    
    # 格式化成Sat Mar 28 22:24:24 2016形式
    print time.strftime("%a %b %d %H:%M:%S %Y", time.localtime()) 
    
    # 将格式字符串转换为时间戳
    a = "Sat Mar 28 22:24:24 2016"
    print time.mktime(time.strptime(a,"%a %b %d %H:%M:%S %Y"))
    #Python 日期和时间 https://www.runoob.com/python/python-date-time.html
    '''


if __name__ == '__main__':
    # 不能设为None,会报错,有些地方需要登陆,所以cookie自己想办法弄吧(并不知道是那个cookie，也不知道怎么写，把自己的cookie弄上去也许号可能会没了)
    #global cookie
    #cookie =""
    try:  # 不用bat启动会报错所以，设计成这样，报错后默认使用单次运行
        if sys.argv[1] == "1":
            print("开始周期运行！")
            starttimer()  # 周期运行
        else:
            usejson()  # 单次运行
    except Exception as err:
        usejson()  # 单次运行
    '''
    # print(datetime.now())
    # get_forum_list_call(my_url)
    #get_threads_from_forum_page(my_url, my_cookie)
    # print(datetime.now())
    '''

get_forum_list.py
import os
import shutil
import json
import hashlib
import signal

from datetime import datetime
from urllib import request as r
from urllib.parse import urlparse
from urllib.parse import unquote
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
#from selenium.webdriver.chrome.options import Options

#path =r'D:\python3.6.8\Scripts\chromedriver.exe'
#chrome_options = Options()

browser = None

def getSource():
    headers = {
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Content-Encoding': 'gzip,deflate,br',
        'Content-Type': 'text/html; charset=utf-8',
        'User-Agent':'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36',
        'Cookie':''  
    }
    #使用copy()防止修改原代码定义dict
    cap = DesiredCapabilities.PHANTOMJS.copy() 
 
    for key, value in headers.items():
        cap['phantomjs.page.customHeaders.{}'.format(key)] = value
 
    # 不载入图片，爬页面速度会快很多
    cap["phantomjs.page.settings.loadImages"] = False
 
    driver = webdriver.PhantomJS(desired_capabilities=cap)#注意selenium的版本，高版本才支持chromedriver.exe.这里selenium==2.53.6
    return driver

# 1.准备根目录
def prepare_home_base_dir(home_base_dir):
    #print(home_base_dir+'model/res')
    #print(home_base_dir + 'res')
    try:
        os.makedirs(home_base_dir)
    except Exception as err:#FileExistsError or OSError:
        print(err)
        print(datetime.now())
        return
        #print('!!!--error--!!! FileExists or OSError return')
        
    shutil.copytree('./model/res',home_base_dir+'res')
    


# 获取单个页面
def get_single_list_page(url):
    browser.get(url)
    return BeautifulSoup(str(browser.page_source), 'lxml')


def get_and_save_src(path, save_path):
    try:
        with r.urlopen(path) as f:
            with open(save_path, 'wb') as wf:
                wf.write(f.read())
    except BaseException:
        print('!!!--error--!!! error on download img src:', path)


def get_info_from_query(url, key):
    query_str = urlparse(url).query
    for query_pair in query_str.split('&'):
        if query_pair.split('=')[0] == key:
            return query_pair.split('=')[1]


# 获取贴吧基本信息,注意不要拿最后一页的地址,因为需要计算楼层数
def get_forum_basic_info(base_url):
    html_tree = get_single_list_page(base_url)
    # with open('model/model_list.html', 'r', encoding='utf-8') as f:
    #     html_tree = BeautifulSoup(f, 'lxml')
        
    res = {}

    # 标题
    res['title'] = html_tree.head.title.string.strip()
    # print(res['title'])

    # 贴吧头像部分
    res['card_head_img'] = html_tree.body.find('img', class_='card_head_img')['src']
    # print(res['card_head_img'])

    card_title = html_tree.body.find('div', class_='card_title')

    # 贴吧名称
    res['card_title_fname'] = card_title.find('a', class_='card_title_fname').string.strip()
    # print(res['card_title_fname'])

    # 关注和帖子数
    res['card_menNum'] = card_title.find('span', class_='card_menNum').string.strip()
    # print(res['card_menNum'])
    res['card_infoNum'] = card_title.find('span', class_='card_infoNum').string.strip()
    # print(res['card_infoNum'])

    # 贴吧口号(目录不要了)
    if html_tree.find('p', class_='card_slogan')!=None:
        res['card_slogan'] = html_tree.find('p', class_='card_slogan').string
    else:
        res['card_slogan'] =''
    # print(res['card_slogan'])

    thread_list_bottom = html_tree.body.find('div', class_='thread_list_bottom')
    # 贴吧页数,这个也很关键
    floor_group = thread_list_bottom.find('div', id='frs_list_pager')
    if floor_group:
        floor_nodes = floor_group.find_all('a')
        last_page = floor_nodes[-1]
        page_link = last_page['href']
        ori_num = int(get_info_from_query(page_link, 'pn'))
        if ori_num == 0:
            res['total_page'] = 1
        else:
            res['total_page'] = int(ori_num / 50 + 1)
    else:
        res['total_page'] = 1
    # print(res['total_page'])

    th_footer_l = thread_list_bottom.find('div', class_='th_footer_l')
    th_footer_l_nodes = th_footer_l.find_all('span')

    # 帖子数
    res['total_thread_num'] = int(th_footer_l_nodes[0].string.strip())
    # print(res['total_thread_num'])
    # 楼层总数
    res['total_floor_num'] = int(th_footer_l_nodes[1].string.strip())
    # print(res['total_floor_num'])
    # 会员数
    res['total_member_num'] = int(th_footer_l_nodes[2].string.strip())
    # print(res['total_member_num'])

    # 贴吧id,从script 源代码中拿
    script_nodes = html_tree.head.find_all('script')
    for script in script_nodes:
        script_str = str(script)
        if script_str.find('PageData.forum') != -1:
            # 切割js源代码,非常恶心
            str1 = script_str[script_str.find('PageData.forum'):-1]
            str2 = str1[0: str1.find('}') + 1]
            str3 = str2[str2.find(':') + 1:str2.find(',')]  # 认为id会写在第一个
            res['fid'] = str3.strip()

    # print(res['fid'])

    return res


def inflate_detail_model_with_list_data(base_info):
    with open('./model/model_list.html', 'r', encoding='utf-8') as f:
        # parser = etree.HTMLParser()
        html_tree = BeautifulSoup(f, 'lxml')

    title = html_tree.head.title
    title.clear()
    title.append(base_info['title'])

    card_head_img = html_tree.body.find('img', class_='card_head_img')
    card_head_img['src'] = base_info['card_head_img']

    card_title = html_tree.body.find('div', class_='card_title')

    card_title_fname = card_title.find('a', class_='card_title_fname')
    card_title_fname.clear()
    card_title_fname.append(base_info['card_title_fname'])

    card_menNum = card_title.find('span', class_='card_menNum')
    card_menNum.clear()
    card_menNum.append(base_info['card_menNum'])

    card_infoNum = card_title.find('span', class_='card_infoNum')
    card_infoNum.clear()
    card_infoNum.append(base_info['card_infoNum'])

    card_slogan = html_tree.find('p', class_='card_slogan')
    card_slogan.clear()
    if base_info['card_slogan']:
        card_slogan.append(base_info['card_slogan'])

    return html_tree


# 列表页不需要登录,所以不用cookie,但是因为有恶心的bigpipe,所以要靠selenium和PhantomJS,这个得提前装好

# 贴吧所有列表页,帖子页分开来弄
def get_forum_list(base_url):
    quote_base_dir = get_info_from_query(base_url, 'kw')
    if quote_base_dir is None:
        print('!!!--error--!!! fatal error, can not find forum name!')
        return
    else:
        base_dir = './'+unquote(unquote(quote_base_dir))
        base_dir += '/'
        #print(base_dir)

    # 1.准备根目录
    print('1.prepare forum folder')
    prepare_home_base_dir(base_dir)

    # 2.收集贴吧基本信息
    print('2.get forum basic info')
    if os.path.exists(base_dir + 'forum_base_info.json'):
        print('2.get forum basic info from json')
        with open(base_dir + 'forum_base_info.json', 'r', encoding='utf-8') as base_info_file:
            info = json.load(base_info_file)

    else:
        print('2.get forum basic info from web')
        info = get_forum_basic_info(base_url)
        with open(base_dir + 'forum_base_info.json', 'w', encoding='utf-8') as base_info_file:
            base_info_file.write(json.dumps(info))

    # print(info)
    # 构造模版,并填充基本数据
    print('3.inflate model with info')
    model = inflate_detail_model_with_list_data(info)

    print('4.start getting pages')
    for page in range(1, info['total_page'] + 1):
        if not os.path.exists(base_dir + str(page) + '.html'):
            print('getting page:' + str(page))
            # 资源目录
            os.makedirs(base_dir + str(page))
            # 清除模版里的旧数据
            thread_list = model.find('ul', id='thread_list')
            thread_list.clear()
            thread_list_bottom = model.find('div', class_='thread_list_bottom')
            thread_list_bottom.clear()
            headpic = model.find('div', class_='card_head')#找到包住吧头像的标签
            headpic2= model.find('img', class_='card_head_img')#先存着吧头像标签
            headpic.clear()#清空包住吧头像的标签
            temp=BeautifulSoup("<a href='"+"http://tieba.baidu.com/f?kw=" + quote_base_dir + "&ie=utf-8"+"'>"+str(headpic2)+"</a>",'lxml')#为吧头像表情增加超链接
            headpic.append(temp)#重新添加回去
            baming = model.find('a', class_='card_title_fname')
            baming['href']="http://tieba.baidu.com/f?kw=" + quote_base_dir + "&ie=utf-8"#点贴吧名字跳转到活的贴吧链接

            print('getting page detail ' + str(page))

            # 获取实际内容
            content = get_single_list_page(
                "http://tieba.baidu.com/f?kw=" + quote_base_dir + "&ie=utf-8&pn=" + str(int((page - 1) * 50)))

            # 帖子列表
            tid_list = []
            thread_list_data = content.find('ul', id='thread_list')

            #username2=content.find('div',class_='threadlist_author')
            #print(username2)
            #username2.clear()

            for thread_list_node in thread_list_data:
                try:
                    if "thread_top_list_folder" in thread_list_node['class']:
                        # 置顶
                        # j_thread_hidden = thread_list_node.find_all('a', class_="j_thread_hidden")
                        # for j_thread_hidden_node in j_thread_hidden:
                        #     tid_str = j_thread_hidden_node['data-field']
                        #     tid_list.append(tid_str[tid_str.find(':') + 1:tid_str.find('}')])

                        j_th_tit = thread_list_node.find_all('a', class_="j_th_tit")
                        if j_th_tit:
                            for j_th_tit_node in j_th_tit:
                                tid_str = ((j_th_tit_node['href']).split('/'))[-1]
                                tid_list.append(tid_str)
                                # 修改帖子指向地址
                                j_th_tit_node['href'] = str(page) + '_thread/' + tid_str + '/1.html'
                        else:
                            print('something wrong maybe')

                        username2=thread_list_node.find_all('div', class_='threadlist_author')#去掉星座图片
                        for temp in username2:
                            if temp:
                                temp.find('span', class_='icon_wrap').clear()

                        thread_list.append(thread_list_node)#主题贴列表相关，注释掉后就仅剩帖子了

                    elif "j_thread_list" in thread_list_node['class']:
                        # 普通帖子
                        # tid_str1 = thread_list_node['data-field']
                        # tid_str2 = tid_str1[tid_str1.find('id') + 3: -1]
                        # tid_str3 = tid_str2[0:tid_str2.find(',')]
                        # tid_list.append(tid_str3.replace(':', '').strip())

                        # 置顶也会被搜进这里来
                        j_th_tit = thread_list_node.find('a', class_="j_th_tit")
                        if j_th_tit:
                            tid_str = j_th_tit['href'].split('/')[-1]
                            tid_list.append(tid_str)
                            # 修改帖子指向地址
                            j_th_tit['href'] = str(page) + '_thread/' + tid_str + '/1.html'
                        else:
                            print('something wrong on a_target maybe')
                            
                        username2=thread_list_node.find('div', class_='threadlist_author').find_all('span', class_='icon_wrap')#去掉星座图片
                        for temp in username2:
                            if temp:
                                temp.clear()
                                
                        thread_list.append(thread_list_node)#主题贴列表相关，注释掉后就仅剩置顶帖了

                    else:
                        # 广告!!!
                        print('enconter an ad ' + str(page))

                    username=thread_list_node.find('span', class_='tb_icon_author').find_all('img', class_='nicknameEmoji')#主题贴楼主昵称图片
                    for temp in username:
                        if username:
                            emoji=temp['src'].split("/")
                            temp['src'] ="res/nickemoji/"+emoji[len(emoji)-1]
                        
                    username=thread_list_node.find('span', class_='tb_icon_author_rely').find_all('img', class_='nicknameEmoji')#最后回贴人的昵称图片
                    for temp in username:
                        if username:
                            emoji=temp['src'].split("/")
                            temp['src'] ="res/nickemoji/"+emoji[len(emoji)-1]

                except BaseException:
                    # 解析帖子列表的时候会混进很多奇怪的东西,不管
                    print('enconter an exception but it is expected')

            # 保存列表中的详情id数据
            thread_id_data = {}
            thread_id_data['fid'] = info['fid']
            thread_id_data['tid_list'] = tid_list
            os.makedirs(base_dir + str(page) + '_thread')
            with open(base_dir + str(page) + '_thread/' + 'tid_info.json', 'w', encoding='utf-8') as tid_info_file:
                tid_info_file.write(json.dumps(thread_id_data))
            
            # 底部页码
            thread_list_bottom_data = content.find('div', class_='thread_list_bottom')
            for pagination_item in thread_list_bottom_data.find_all('a', class_='pagination-item'):
                pagination_item['href'] = None
                if pagination_item.string == '首页':
                    pagination_item['href'] = '1.html'
                elif pagination_item.string == '尾页':
                    pagination_item['href'] = str(info['total_page']) + '.html'
                elif pagination_item.string.find('上一页') != -1:
                    pagination_item['href'] = str(page - 1) + '.html'
                elif pagination_item.string.find('下一页') != -1:
                    pagination_item['href'] = str(page + 1) + '.html'
                else:
                    try:
                        num = int(pagination_item.string)
                        pagination_item['href'] = str(num) + '.html'
                    except TypeError:
                        print('!!!--error--!!! TypeError:' + pagination_item.prettify())
                    except ValueError:
                        print('!!!--error--!!! unknow page:' + pagination_item.string)

            thread_list_bottom.replace_with(thread_list_bottom_data)

            # 处理资源,这里再弄一份是因为model突然搜不到带src的标签,我也不知道为什么,重新构造一份才行
            write_model = BeautifulSoup(str(model), 'lxml')
 
            src_nodes = write_model.select('img')
            print('dealing with src ' + str(page))
            count = 1
            download_list = {}
            for src_node in src_nodes:
                # 有些大图放在bpic里面,直接去下大图太大了。。。还是要data-original这个是小图
                try:
                    if src_node['src'] is None or src_node['src'] == '':
                        src_node['src'] = src_node['data-original']
                        src_node['style'] = 'display: inline; width: 90px; height: 90px;'
                    src_str = src_node['src']
                    if src_str.find('http') != -1:
                        if download_list.get(src_str):
                            src_node['src'] = download_list.get(src_str)
                        else:
                            md5 = hashlib.md5()
                            md5.update(src_str.encode('utf-8'))
                            link_path = str(page) + '/' + md5.hexdigest() + '.png'
                            src_node['src'] = link_path
                            get_and_save_src(src_str, base_dir + link_path)
                            print('downloading no.' + str(count) + ' img ' + str(page))
                            count += 1
                            download_list[src_str] = link_path
                except BaseException:
                    print('!!!--error--!!! invalid img tag!')
            download_list.clear()
                
            # 完成,输出
            with open(base_dir + str(page) + '.html', 'w', encoding='utf-8') as f:
                f.write(str(write_model))
            print('done with page ' + str(page))


if __name__ == '__main__':
    #browser = webdriver.Chrome()
    browser = getSource()#webdriver.PhantomJS()
    get_forum_list("https://tieba.baidu.com/f?ie=utf-8&kw=%E8%B4%B4%E5%90%A7")#贴吧
    browser.service.process.send_signal(signal.SIGTERM)
    browser.quit()
